File: Leetcode/1577/1577.go
func solve(A []int, B []int) int {
	unmap := make(map[int64]int)
	for i := 0; i < len(B)-1; i++ {
		for j := i+1; j < len(B); j++ {
			unmap[int64(B[i])*int64(B[j])]++
		}
	}
	ans := 0
	for _, v := range A {
		val, ok := unmap[int64(v)*int64(v)]
		if ok {
			ans += val
		}
	}
	return ans
}
func numTriplets(nums1 []int, nums2 []int) int {
	return solve(nums1, nums2) + solve(nums2, nums1)
}

File: Leetcode/1577/1577.cpp
using ll = long long;

class Solution {
public:
    int solve(vector<int>& nums1, vector<int>& nums2) {
        int ans = 0;
        unordered_map<ll, int> unmap;
        int n = (int)nums2.size();
        for (int j = 0; j < n - 1; ++j) {
            for (int k = j+1; k < n; ++k) {
                ++unmap[(ll)nums2[j]*nums2[k]];
            }
        }
        ll temp;
        for (const auto &i : nums1) {
            temp = (ll)i*i;
            auto iter = unmap.find(temp);
            if (iter != unmap.end()) {
                ans += iter->second;
            }
        }
        return ans;
    }
    int numTriplets(vector<int>& nums1, vector<int>& nums2) {
        return solve(nums1,nums2)+solve(nums2,nums1);
    }
};